### PR TITLE
Feat/ts execution tests

### PR DESCRIPTION
**Pull request by rkodev**
Resolves https://github.com/microsoftgraph/msgraph-sdk-raptor/issues/1041

Refactors TS tests to externalize dependencies (should take advantage of dependabot)
Adds execution tests for TS
Removes indirection in the flow of TS tests execution
minor: Improve code execution time, avoid parsing the sources twice

### CODE REVIEW

**Review by timayabi2020 — APPROVED** [review]

File: TestsCommon/TypeScriptTestRunner.cs [hunk 1]
```diff
@@ -1,339 +1,345 @@
...
+            ArgumentNullException.ThrowIfNull(compilationDirectory);
+            ArgumentNullException.ThrowIfNull(languageTestData);
 
+            foreach (var testData in languageTestData)
+            {
+                var fullPath = Path.Join(GraphDocsDirectory.GetSnippetsDirectory(testData.Version, Languages.TypeScript), testData.FileName);
+                Assert.IsTrue(File.Exists(fullPath), "Snippet file referenced in documentation is not found!");
 
-            var codeSnippetFormatted = match.Groups[1].Value;
+                var fileContent = File.ReadAllText(fullPath);
```
Comment (zengin) [inline TestsCommon/TypeScriptTestRunner.cs:115]: please use the Async alternative.

File: TestsCommon/TypeScriptTestRunner.cs [hunk 2]
```diff
@@ -1,339 +1,345 @@
...
-            var BuildPath = Path.Combine(RootPath, "build");
-            return await ExecuteNPMProcess(BuildPath, "tsc", "-p tsconfig.json --outDir ./build", 600, false).ConfigureAwait(false);
-        }
+                    this.NpmResults[item.Key] = payload;
+                }
 
+                foreach (var item in errors)
+                    File.Move(Path.Combine(buildDir, item.Key), Path.Combine(backUpDir, item.Key));
 
+                if (errors.Count == 0) break;
```
Comment (zengin) [inline TestsCommon/TypeScriptTestRunner.cs:190]: nit: run this check right after parsing NPM errors.

File: TestsCommon/TypeScriptTestRunner.cs [hunk 3]
```diff
@@ -1,339 +1,345 @@
...
-        private async static Task<(string, string)> CompileTypescriptFiles(string RootPath)
-        {
-            var BuildPath = Path.Combine(RootPath, "build");
-            return await ExecuteNPMProcess(BuildPath, "tsc", "-p tsconfig.json --outDir ./build", 600, false).ConfigureAwait(false);
-        }
+                    this.NpmResults[item.Key] = payload;
+                }
 
+                foreach (var item in errors)
+                    File.Move(Path.Combine(buildDir, item.Key), Path.Combine(backUpDir, item.Key));
```
Comment (zengin) [inline TestsCommon/TypeScriptTestRunner.cs:188]: Do we need a separate loop here? Can we make File.Move part of the loop above?

File: TestsCommon/TypeScriptTestRunner.cs [hunk 4]
```diff
@@ -1,339 +1,345 @@
...
+            {
+                Assert.Pass();
             }
+            else
+            {
+                var diagnostic = this.NpmResults[fileName];
 
-            return result;
+                var compilationResultsModel = new CompilationResultsModel(false, diagnostic, testData.FileName);
+                var fileContent = File.ReadAllText(Path.Combine(backupDirectory, fileName));
```
Comment (zengin) [inline TestsCommon/TypeScriptTestRunner.cs:269]:
```suggestion
                var fileContent = await File.ReadAllTextAsync(Path.Combine(backupDirectory, fileName)).ConfigureAwait(false);
```

File: TestsCommon/TypeScriptTestRunner.cs [hunk 5]
```diff
@@ -1,339 +1,345 @@
...
+
+                var diagnostic = new List<Diagnostic>();
+                foreach (string line in lines)
+                {
 
-                    result[fileName] = payload;
+                    var errorPositionsMatch = TSErrorPositions.Match(line);
+                    if (errorPositionsMatch.Success)
+                    {
+                        var errorPositionStart = int.Parse(errorPositionsMatch.Groups[2].Value);
```
Comment (zengin) [inline TestsCommon/TypeScriptTestRunner.cs:317]: please address the warnings raised in the build process.

File: TestsCommon/TypeScriptTestRunner.cs [hunk 6]
```diff
@@ -1,339 +1,345 @@
...
+                                                                                        new LinePosition(errorPositionStart, 0),
+                                                                                        new LinePosition(errorPositionStart, errorPositionEnd))))
+                            );
+                    }
                 }
-            }
 
-            return result;
+                var compilationResultsModel = new CompilationResultsModel(false, diagnostic, testData.FileName);
+                var fileContent = File.ReadAllText(fileLocation);
```
Comment (zengin) [inline TestsCommon/TypeScriptTestRunner.cs:338]:
```suggestion
                var fileContent = await File.ReadAllTextAsync(fileLocation).ConfigureAwait(false);
```

File: TypeScriptV1Tests/SnippetCompileV1Tests.cs [hunk 7]
```diff
@@ -10,56 +10,45 @@ namespace TypeScriptV1Tests
...
-
-        private static RunSettings TestRunSettings => new RunSettings(
+        private static RunSettings runSettings => new RunSettings(
             TestContext.Parameters,
             new RunSettings()
             {
                 Version = Versions.V1,
                 Language = Languages.TypeScript,
-                TestType = TestType.CompilationStable
+                TestType = TestType.ExecutionStable
```
Comment (zengin) [inline TypeScriptV1Tests/SnippetCompileV1Tests.cs:19]: I don't think this change is intended.

File: TestsCommon/MicrosoftGraphTypescriptCompiler.cs [hunk 8]
```diff
@@ -1,47 +0,0 @@
-﻿using System.Collections.ObjectModel;
```
Comment (zengin) [inline TestsCommon/MicrosoftGraphTypescriptCompiler.cs:1]: Thank you for removing this layer!

**Review by zengin — CHANGES_REQUESTED** [review]

File: TestsCommon/TypeScriptTestRunner.cs [hunk 9]
```diff
@@ -1,339 +1,345 @@
...
-        public static void RunTest(string path, LanguageTestData testData, Dictionary<string, Collection<Dictionary<string, string>>> npmResults)
+        private static async Task prepareNPM(string compilationDirectory)
         {
-            ArgumentNullException.ThrowIfNull(npmResults);
-            ArgumentNullException.ThrowIfNull(testData);
-
-#pragma warning disable CA1308 // Normalize strings to uppercase
-            var fileName = $"{FILE_PREFIX}-{testData.FileName.ToLowerInvariant().Replace(" ", "-")}.ts";
-#pragma warning restore CA1308 // Normalize strings to uppercase
+            DirectoryInfo directoryInfo = new DirectoryInfo(Path.Combine(Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location), "TsDependecies"));
```
[the remark this answers is not in the document]
Comment (rkodev) [reply TestsCommon/TypeScriptTestRunner.cs:81]: Yes the `TsDependencies` are copied over from the project folder `TypeScriptV1Tests` for all the other tests. For example we can know issues project references the dependencies via the project import [here ](https://github.com/microsoftgraph/msgraph-sdk-raptor/blob/feat/ts-execution-tests/TypeScriptV1KnownIssueTests/TypeScriptV1KnownIssueTests.csproj)


```
  <ItemGroup>
    <None Remove="..\TypeScriptV1Tests\TsDependecies\package-lock.json" />
    <None Remove="..\TypeScriptV1Tests\TsDependecies\package.json" />
    <None Remove="..\TypeScriptV1Tests\TsDependecies\tsconfig.json" />
  </ItemGroup>

  <ItemGroup>
    <Content Include="..\TypeScriptV1Tests\TsDependecies\package-lock.json">
      <CopyToOutputDirectory>Always</CopyToOutputDirectory>
    </Content>
    <Content Include="..\TypeScriptV1Tests\TsDependecies\package.json">
      <CopyToOutputDirectory>Always</CopyToOutputDirectory>
    </Content>
    <Content Include="..\TypeScriptV1Tests\TsDependecies\tsconfig.json">
      <CopyToOutputDirectory>Always</CopyToOutputDirectory>
    </Content>
  </ItemGroup>

```

**Review by zengin — APPROVED** [review]

**Pull request merged**